### DISCUSSION TITLE
Types for custom endpoints

### DISCUSSION THREAD

**Discussion by rhyst** [directus/directus#8552] [Archive]
Is it possible to import the types for the custom endpoint parameters? 

```
import { Router } from 'express';
import { SomeType } from 'somedirectusthing';

function registerEndpoint(router, { services: Router, exceptions }: SomeType) {
...
}
```

At the moment I need to read through the source code to try to work out what services are available to me.

Also is there more documentation on extensions? There's more for the JS SDK but it doesn't quite match up with what is provided to the endpoint function. Is it possible to just use the SDK in an extension?

Thanks for any help.

**Comment by nickrum** [comment]
Yes, you can import a `defineEndpoint` helper function from `@directus/extensions-sdk` and do something like this:

```ts
import { defineEndpoint } from '@directus/extensions-sdk';

export default defineEndpoint((router) => {
  ...
});
```

> Also is there more documentation on extensions?

We are working on improving the extension docs.

> Is it possible to just use the SDK in an extension?

You can certainly do that, but it is faster and more efficient to access the services directly.

**Comment by rhyst** [comment]
Ah thank you, that's great. Are there types anywhere for the `ApiExtensionContext.services` yet? At the moment it seems like it is typed as `any` in that `defineEndpoint` function

**Reply by nickrum** [reply]
Not yet, unfortunately. The properties of `services` are classes, so we'd have to duplicate their type signatures into the `@directus/shared` package to be able to expose them through `defineEndpoint`. This will be done at some point. For now you could import the classes from the `directus` package directly.
